### PR TITLE
type hints from monkeytype

### PR DESCRIPTION
**Pull request by dckc**
`mypy --strict *.py src/*.py` is happy (`mypy --version` = 0.560)

The change comments aren't pretty (I don't have my usual magit setup in my freebsd VM) I hope to clean them up.

### CODE REVIEW

**Comment by dckc** [comment]
Testing shows `_Hash` is only in the stubs for `hashlib`. Ugh. The 49a42c6 fix works at runtime, but to check the types use: `mypy --ignore-missing-imports src/*.py`.

File: src/builder.py [hunk 1]
```diff
@@ -41,7 +44,7 @@ def __exit__(self, type, value, traceback):
 
 
 class FileHandle:
-    def __init__(self, builder, path):
+    def __init__(self, builder: 'Builder', path: str) -> None:
```
Comment (EdSchouten) [inline src/builder.py:49]: I have to confess I'm not an expert on Python's type hinting: what's with the string quoted type name?

File: src/builder.py [hunk 2]
```diff
@@ -130,12 +133,12 @@ def host(self):
     def rename(self, dst):
         os.rename(self._path, dst._path)
 
-    def cmake(self, args=[]):
+    def cmake(self, args: List[Any] = []) -> 'FileHandle':
```
Comment (EdSchouten) [inline src/builder.py]: This should be `List[str]`, right?

File: src/distfile.py [hunk 3]
```diff
@@ -63,7 +65,7 @@ def _apply_patch(patch, target):
                     break
 
         # Apply the patch.
-        with open(patch) as f:
+        with open(patch, 'rb') as f:
```
Comment (EdSchouten) [inline src/distfile.py:69]: Is this change trying to fix some specific bug/regression?

File: src/util.py [hunk 4]
```diff
@@ -4,16 +4,20 @@
 
 import gzip
 import hashlib
+from _hashlib import HASH as _Hash
```
Comment (EdSchouten) [inline src/util.py]: What about leaving functions that use this type untyped? This looks like an implementation detail on which I think we shouldn't depend.

**Review by EdSchouten — COMMENTED** [review]
Thanks! I think this is worth pursuing!

File: src/builder.py (hunk 1 above)
Comment (dckc) [reply src/builder.py:49]: It's a [kludge for handling forward references](https://www.python.org/dev/peps/pep-0484/#forward-references).

File: src/builder.py (hunk 2 above)
Comment (dckc) [reply src/builder.py]: yes, I suppose so.

monkeytype didn't figure that out and `List[Any]` didn't cause type checking to fail, so I didn't notice it.

File: src/distfile.py (hunk 3 above)
Comment (dckc) [reply src/distfile.py:69]: I think so, but I rushed the commit messages and of course now I don't remember clearly.

Perhaps it's not actually necessary... perhaps it's left over from flailing around. When a variable like f gets used as more than one type in the same function, mypy's diagnostics are bad and I was pulling my hair out trying to figure out what was going on.

File: src/util.py (hunk 4 above)
Comment (dckc) [reply src/util.py]: Yes, you're probably right.

File: src/distfile.py (hunk 3 above)
Comment (dckc) [reply src/distfile.py:69]: Ah... right... earlier in that function, line 47, we have `with open(patch, 'rb') as f:` so f has type `BinaryIO`. `open(patch)` returns `TextIO`. Without this `rb` arg, we get:

```
src/distfile.py:68: error: Incompatible types in assignment (expression has type "TextIO", variable has type "BinaryIO")
```

File: src/builder.py [hunk 5]
```diff
@@ -276,8 +279,26 @@ def get_new_executable(self):
         return path
 
 
-class HostBuilder:
-    def __init__(self, build_directory, install_directory):
+class Abstract(NotImplementedError):
+    pass  # TODO: use abc?
+
+
+class Builder:
```
Comment (EdSchouten) [inline src/builder.py]: Would it make sense to use https://docs.python.org/3/library/abc.html here?

File: src/builder.py [hunk 6]
```diff
@@ -276,8 +279,26 @@ def get_new_executable(self):
...
+    def __init__(self, build_directory: BuildDirectory,
+                 install_directory: Optional[str]) -> None:
```
Comment (EdSchouten) [inline src/builder.py]: The constructor arguments serve no purpose, right? Maybe let this be a memberless class?

File: src/builder.py [hunk 7]
```diff
@@ -276,8 +279,26 @@ def get_new_executable(self):
...
+        self._build_directory = build_directory
+        self._install_directory = install_directory
+
+    def cmake(self, builddir: str, srcdir: str,
+              args: List[str]) -> None: raise Abstract()
+
+    def install(self, source: str, target: str) -> None: raise Abstract()
+
+    def run(self, cwd: str, command: List[str]) -> None: raise Abstract()
```
Comment (EdSchouten) [inline src/builder.py]: Instead of raising `Abstract`, we can probably use `@abstractmethod` here.

File: src/distfile.py [hunk 8]
```diff
@@ -8,16 +8,18 @@
...
 
+from typing import Set, Union
 log = logging.getLogger(__name__)
 
 
 class Distfile:
-    def __init__(self, distdir, name, checksum, master_sites, patches,
-                 unsafe_string_sources):
+    def __init__(self, distdir: str, name: str, checksum: str, master_sites: Set[str], patches: Set[str],
+                 unsafe_string_sources: Union[frozenset, Set[str]]) -> None:
```
Comment (EdSchouten) [inline src/distfile.py]: Is the `Union[frozenset, Set[str]]` really necessary? As in, wouldn't `Set[str]` be sufficient?

File: src/distfile.py (hunk 3 above)
Comment (EdSchouten) [reply src/distfile.py:69]: Sounds all right then!

File: src/package.py [hunk 9]
```diff
@@ -11,13 +11,16 @@
...
+from typing import AbstractSet, Any, Callable, Dict, Optional, Set, Union
 log = logging.getLogger(__name__)
 
 
 class HostPackage:
-    def __init__(self, install_directory, name, version, homepage,
-                 build_depends, lib_depends, distfiles, build_cmd,
-                 resource_directory):
+    def __init__(self, install_directory: str, name: str, version: SimpleVersion, homepage: str,
+                 build_depends: Set['HostPackage'], lib_depends: Set['HostPackage'], distfiles: Dict[str, Distfile], build_cmd: Callable,
```
Comment (EdSchouten) [inline src/package.py]: Out of curiosity, there is no way to specify the typing of the callable?

File: src/rpm.py [hunk 10]
```diff
@@ -5,13 +5,14 @@
 import struct
 
 
+from typing import Any, Dict, Iterator, Iterable, List, Union
 class Header:
     """Class for generating binary RPM headers."""
 
-    def __init__(self, entries):
+    def __init__(self, entries: Dict[int, Any]) -> None:
```
Comment (EdSchouten) [inline src/rpm.py]: I suspect the `Any` can be prevented here by adding a proper entry base type?

**Review by EdSchouten — COMMENTED** [review]
Thanks for working on this. This is starting to look pretty awesome!

In addition to the comments I gave, could you also run `./reformat.sh` as well? It's a script that automatically invokes yapf on all Python source files in the tree.

File: src/builder.py (hunk 5 above)
Comment (dckc) [reply src/builder.py]: yes... hence the TODO.

1st time I tried abc with mypy, mypy didn't grok. But that was a while ago. Let's see if it works now...

File: src/builder.py (hunk 6 above)
Comment (dckc) [reply src/builder.py]: They serve to define the types of `self._build_directory` and `self._install_directory` which are referred to elsewhere.

I made `Builder` inherit from `ABC` and marked the other methods `@abstractmethod` so it is a memberless class now.

File: src/builder.py (hunk 7 above)
Comment (dckc) [reply src/builder.py]: yes...

File: src/distfile.py (hunk 8 above)
Comment (dckc) [reply src/distfile.py]: I think `frozenset()` is passed in for this argument somewhere, and `Set[str]` doesn't include frozen sets. `AbstractSet` includes both. I wish the shorter name were the immutable type, but this isn't scala...

File: src/package.py (hunk 9 above)
Comment (dckc) [reply src/package.py]: yes, there is... `Callable[args, return type]` or something like that...

File: src/rpm.py (hunk 10 above)
Comment (dckc) [reply src/rpm.py]: yes; I think a `Union` is in order here.

File: packages/boost/BUILD [hunk 11]
```diff
@@ -12,24 +12,25 @@ def build(ctx):
...
-        '--prefix=%s' % stage,
-        'install',
-    ] + {
-        'armv6': ['abi=aapcs', 'address-model=32', 'architecture=arm'],
-        'armv7': ['abi=aapcs', 'address-model=32', 'architecture=arm'],
-        'aarch64': ['abi=aapcs', 'address-model=64', 'architecture=arm'],
-        'i686': ['abi=sysv', 'address-model=32', 'architecture=x86'],
-        'x86_64': ['abi=sysv', 'address-model=64', 'architecture=x86'],
-    }[ctx.cpu()])
+    srcdir.run(
```
Comment (EdSchouten) [inline packages/boost/BUILD:15]: It looks like you're use YAPF 0.20, whereas I was still using 0.19. Just upgraded YAPF on my system and got the same output. Fair enough!

File: src/catalog.py [hunk 12]
```diff
@@ -270,14 +289,15 @@ def package(self, package, version):
 
 
 class FreeBSDCatalog(Catalog):
-    def __init__(self, old_path, new_path):
+    def __init__(self, old_path: None, new_path: str) -> None:
```
Comment (EdSchouten) [inline src/catalog.py]: `None` should be `Union[None, str]`, right?

File: src/catalog.py [hunk 13]
```diff
@@ -529,19 +553,20 @@ def package(self, package, version):
 
 
 class NetBSDCatalog(Catalog):
-    def __init__(self, old_path, new_path):
+    def __init__(self, old_path: None, new_path: str) -> None:
```
Comment (EdSchouten) [inline src/catalog.py]: Same as above. Probably in more places below?

**Review by EdSchouten — COMMENTED** [review]
Sorry for letting you go through so many rounds of reviews, but I think we're *almost* good to go. Be sure to fix that tiny issue and I'll get this merged for you.

Thanks!

File: src/catalog.py (hunk 12 above)
Comment (dckc) [reply src/catalog.py]: aka `Optional[str]`, yes.

File: src/catalog.py (hunk 13 above)
Comment (dckc) [reply src/catalog.py]: I wonder if adding`build_packages.py` to the files to check would catch things like this.

Comment (dckc) [reply src/catalog.py]: By using `mypy --strict` I found a whole bunch of stuff that should be fixed. Stand by...

**Comment by dckc** [comment]
@EdSchouten I think I'm done now.

`mypy --strict *.py src/*.py` is happy.

**Comment by EdSchouten** [comment]
Thanks a lot, Dan!

**Pull request merged**